dronegen/misc.go

```diff
@@ -20,9 +20,74 @@ import (
+	dockerPipeline.Workspace = workspace{Path: "/go"}
+
+	// Add docker service
+	dockerPipeline.Services = []service{
+		dockerService(),
+	}
+	dockerPipeline.Volumes = dockerVolumes()
+
+	dockerPipeline.Steps = append(dockerPipeline.Steps, verifyTaggedBuildStep())
+	dockerPipeline.Steps = append(dockerPipeline.Steps, waitForDockerStep())
+
+	// Pull/Push Steps
+	dockerPipeline.Steps = append(dockerPipeline.Steps, step{
+		Name:  "Pull/retag Docker images",
+		Image: "docker",
+		Environment: map[string]value{
+			"AWS_ACCESS_KEY_ID":          {fromSecret: "TELEPORT_BUILD_USER_KEY"},
+			"AWS_SECRET_ACCESS_KEY":      {fromSecret: "TELEPORT_BUILD_USER_SECRET"},
+			"DOCKER_PRODUCTION_USERNAME": {fromSecret: "PRODUCTION_QUAYIO_DOCKER_USERNAME"},
+			"DOCKER_PRODUCTION_PASSWORD": {fromSecret: "PRODUCTION_QUAYIO_DOCKER_PASSWORD"},
+		},
+		Volumes: dockerVolumeRefs(),
+		Commands: []string{
+			"apk add --no-cache aws-cli",
+			"export VERSION=${DRONE_TAG##v}",
+			// authenticate with staging credentials
+			"aws ecr get-login-password --region=us-west-2 | docker login -u=\"AWS\" --password-stdin 146628656107.dkr.ecr.us-west-2.amazonaws.com",
+			// pull 'temporary' CI-built images
+			"echo \"---> Pulling images for $${VERSION}\"",
+			"docker pull 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport:$${VERSION}",
+			"docker pull 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport-ent:$${VERSION}",
+			"docker pull 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport-ent:$${VERSION}-fips",
+			// retag images to production naming
+			"echo \"---> Tagging images for $${VERSION}\"",
+			"docker tag 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport:$${VERSION} quay.io/gravitational/teleport:$${VERSION}",
+			"docker tag 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport-ent:$${VERSION} quay.io/gravitational/teleport-ent:$${VERSION}",
+			"docker tag 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport-ent:$${VERSION}-fips quay.io/gravitational/teleport-ent:$${VERSION}-fips",
+			//retag ECR images
+			"docker tag 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport:$${VERSION} public.ecr.aws/gravitational/teleport:$${VERSION}",
+			"docker tag 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport-ent:$${VERSION} public.ecr.aws/gravitational/teleport-ent:$${VERSION}",
+			"docker tag 146628656107.dkr.ecr.us-west-2.amazonaws.com/gravitational/teleport-ent:$${VERSION}-fips public.ecr.aws/gravitational/teleport-ent:$${VERSION}-fips",
+			// authenticate with production credentials
+			"docker logout 146628656107.dkr.ecr.us-west-2.amazonaws.com",
+			"docker login -u=\"$DOCKER_PRODUCTION_USERNAME\" -p=\"$DOCKER_PRODUCTION_PASSWORD\" quay.io",
+			// push production images
+			"echo \"---> Pushing images for $${VERSION}\"",
+			"docker push quay.io/gravitational/teleport:$${VERSION}",
```

Comment:
If quay is down and this line fails we won't publish to the main ECR site.  I think we should have promote to ECR and promote to quay in different pipelines, such that neither will block the other.

Reply:
Updated for multiple pipelines in https://github.com/gravitational/teleport/pull/14347/commits/a44b732aff01b95ceb6e55485de71970efdd0808